scipy/cluster/_heap.pxi

```diff
@@ -0,0 +1,73 @@
+# cython: boundscheck=False, wraparound=False, cdivision=True
+import numpy as np
+
+
+cdef class Heap:
+    cdef int[:] index_by_key
+    cdef int[:] key_by_index
+    cdef double[:] values
+    cdef int size
+
+    def __init__(self, double[:] values):
+        self.size = values.shape[0]
+        self.index_by_key = np.arange(self.size, dtype=np.intc)
+        self.key_by_index = np.arange(self.size, dtype=np.intc)
+        self.values = values.copy()
+        cdef int i
+        for i in reversed(range(self.size / 2)):
+            self.sift_down(i)
+
+    cdef tuple get_min(self):
```

Comment:
I was aware of that. I think you are right, just split this method into two, if we are trying to optimize for speed.